### ISSUE TITLE
Class name does not get a coloring in constructor call

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#501]
In the following code:

```c++
struct Waldo {
  Waldo();
  Waldo(const Waldo&);
};

char Helper(...);

const int x = sizeof(Helper(Waldo()));  // <--
```

The `Waldo` token on the indicator line does not get a coloring.

If `struct Waldo`'s copy constructor is removed, the coloring appears.

**Comment by HighCommander4** [comment]
Go-to-def does not work on that token either.

**Comment by HighCommander4** [comment]
I suspect this has to do with the fact that `Waldo` cannot be passed to a variadic function. However, this is a runtime restriction; in an unevaluated context such as the argumen to `sizeof`, it appears to be valid.

**Comment by HighCommander4** [comment]
Note that this came up in the context of gtest. Here is a less-reduced example:

```c++
#include "gtest/gtest.h"

struct Waldo {
  Waldo();
  Waldo(const Waldo&);
  bool operator==(const Waldo&) const;
};

void foo() { EXPECT_EQ(Waldo(), Waldo()); }  // <--
```

Here, the first `Waldo` token on the indicated line is not colored.

**Comment by sam-mccall** [comment]
This is really curious! If you hover on the token, you get "function __builtin_trap"!

The AST for the sizeof looks like 
```
UnaryExprOrTypeTraitExpr 0x57c7fc995d80 </usr/local/google/home/sammccall/test.cc:7:15, col:37> 'unsigned long' sizeof
`-ParenExpr 0x57c7fc995d60 <col:21, col:37> 'char'
  `-CallExpr 0x57c7fc995a90 <col:22, col:36> 'char'
    |-ImplicitCastExpr 0x57c7fc995a78 <col:22> 'char (*)(...)' <FunctionToPointerDecay>
    | `-DeclRefExpr 0x57c7fc995a30 <col:22> 'char (...)' lvalue Function 0x57c7fc995700 'Helper' 'char (...)' non_odr_use_unevaluated
    `-BinaryOperator 0x57c7fc995d40 <col:29, col:35> 'Waldo' ','
      |-CallExpr 0x57c7fc995d20 <col:29, col:35> 'void'
      | `-ImplicitCastExpr 0x57c7fc995d08 <col:29> 'void (*)() __attribute__((noreturn)) noexcept' <BuiltinFnToFnPtr>
      |   `-DeclRefExpr 0x57c7fc995c88 <col:29> '<builtin fn type>' Function 0x57c7fc995b68 '__builtin_trap' 'void () __attribute__((noreturn)) noexcept' non_odr_use_unevaluated
      `-CXXTemporaryObjectExpr 0x57c7fc9958d8 <col:29, col:35> 'Waldo' 'void ()'
```

which is to say, the arg to Helper is actually the comma-expression `__builtin_trap(), Waldo()`. And lots of these nodes land on the "Waldo" token.

Our selection tree here is:

```
TranslationUnitDecl 
   VarDecl const int x = sizeof (Helper(__builtin_trap() , Waldo()))
     ImplicitCastExpr sizeof (Helper(__builtin_trap() , Waldo()))
       UnaryExprOrTypeTraitExpr sizeof (Helper(__builtin_trap() , Waldo()))
         ParenExpr (Helper(__builtin_trap() , Waldo()))
           CallExpr Helper(__builtin_trap() , Waldo())
             BinaryOperator __builtin_trap() , Waldo()
               CallExpr __builtin_trap()
                 ImplicitCastExpr __builtin_trap
                  .DeclRefExpr __builtin_trap
```

i.e. *only* __builtin_trap is selected, because the two of them both claim ownership of the `Waldo` token and as siblings, it's first-come-first-served.

---

In summary:
 - something's inserting a call to builtin_trap, presumably to communicate the broken-at-runtime semantics you're talking about
 - there's no "implicit" bit on expressions, and it has a valid location, so we see it as a real node
 - for semantic highlighting, we generate two conflicting tokens and drop both
 - for other operations, Selection picks the wrong target
